File: FoosballGames.Infrastructure/Messaging/ICommandQueryDispatcher.cs
namespace FoosballGames.Infrastructure.Messaging;

public interface ICommandQueryDispatcher
{
    Task SendAsync<TCommand>(TCommand command) where TCommand : ICommand;
    Task<TResponse> SendAsync<TQuery, TResponse>(TQuery query) where TQuery : IQuery<TResponse>;
}

public class CommandQueryDispatcher : ICommandQueryDispatcher
{
    private readonly Func<Type, object> _serviceProvider;

    private static readonly Type CommandHandlerType = typeof(ICommandHandler<>);

    private static readonly Type QueryHandlerType = typeof(IQueryHandler<,>);

    public CommandQueryDispatcher(Func<Type, object> serviceProvider)
    {
        _serviceProvider = serviceProvider;
    }

    public Task SendAsync<TCommand>(TCommand command) where TCommand : ICommand
    {
        var commandType = command.GetType();
        var genericHandlerType = CommandHandlerType.MakeGenericType(commandType);

        var handler = _serviceProvider(genericHandlerType);
        return ((ICommandHandler<TCommand>) handler).HandleAsync(command);
    }

    public Task<TResponse> SendAsync<TQuery, TResponse>(TQuery query) where TQuery : IQuery<TResponse>
    {
        var queryType = query.GetType();
        var responseType = typeof(TResponse);
        var genericHandlerType = QueryHandlerType.MakeGenericType(queryType, responseType);
        var handler = _serviceProvider(genericHandlerType);

        return ((IQueryHandler<TQuery, TResponse>) handler).HandleAsync(query);
    }
}

File: FoosballGames.Infrastructure/Messaging/ICommandQueryDispatcher.fs
namespace FoosballGames.Infrastructure.Messaging

open System
open System.Threading.Tasks
open FoosballGames.Infrastructure.Messaging

type ICommandQueryDispatcher =
    abstract member SendAsync: command: 'TCommand -> Task when 'TCommand :> ICommand
    abstract member SendAsync: query: 'TQuery -> Task<'TResponse> when 'TQuery :> IQuery<'TResponse>
    
type CommandQueryDispatcher(serviceProvider: Type -> obj) =
    let _serviceProvider = serviceProvider
    static let CommandHandlerType = typedefof<ICommandHandler<_>>
    static let QueryHandlerType = typedefof<IQueryHandler<_,_>>
    
    interface ICommandQueryDispatcher with
        member this.SendAsync<'TCommand when 'TCommand :> ICommand>(command: 'TCommand): Task =
            let commandType = command.GetType()
            let genericHandlerType = CommandHandlerType.MakeGenericType(commandType)
            let handler = _serviceProvider(genericHandlerType)
            (handler :?> ICommandHandler<'TCommand>).HandleAsync(command)

        member this.SendAsync<'TQuery, 'TResponse when 'TQuery :> IQuery<'TResponse>>(query: 'TQuery): Task<'TResponse> =
            let queryType = query.GetType();
            let responseType = typedefof<'TResponse>;
            let genericHandlerType = QueryHandlerType.MakeGenericType(queryType, responseType);
            let handler = _serviceProvider(genericHandlerType);
            (handler :?> IQueryHandler<'TQuery, 'TResponse>).HandleAsync(query);